### PR TITLE
Init CD

### PR DESCRIPTION
**Pull request by SarahJaine**
**Description of change**
In short, this code adds an automated deployment pipeline.
- Add deploy job to circle ci config
- Call deploy job only on deployment branches (adhoc fork `main` for dev, HHS `main` for staging, HHS `prod` for prod)
- Change frontend `yarn build` command to change build output dir, tell `index.js` about this new dir location
- Add backend `yarn build` command to create compiled copy of BE files, changes `yarn start` to use new compiled index.js

**NOTE: Before this code is merged to the upstream HHS/Head-Start-TTADP repo, the HHS project circle ci env variables must be added for govcloud's deployer username/password and space name in order for the staging deploy to be successful.  The exact variable names should be:**
- CLOUDGOV_STAGING_USERNAME
- CLOUDGOV_STAGING_PASSWORD
- CLOUDGOV_STAGING_SPACE

**How to test**
Check out latest build for this branch on circle ci https://app.circleci.com/pipelines/github/adhocteam/Head-Start-TTADP/230/workflows/6b5cf689-4975-4e07-b9cf-054a9e4ff23b

**Issue(s)**
* Closes https://github.com/HHS/Head-Start-TTADP/issues/12

**Checklist**

- [x] Meets issue criteria
- [x] Code tested
- [n/a] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [x] Documentation updated
    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -102,3 +215,14 @@ workflows:
...
+          requires:
+            - lint
+            - test_backend
+            - test_frontend
+          filters:
+            branches:
+              only:
+                - << pipeline.parameters.dev_git_branch >>
+                - << pipeline.parameters.staging_git_branch >>
+                - << pipeline.parameters.prod_git_branch >>
```
Comment (SarahJaine) [inline .circleci/config.yml:228]: FWIW I'm not completely pleased with this filtering; it's not exactly skip/fail fast. This limits the deploy job to any branch named `main` or `production`.  In the deploy job, after assets are built and cf is installed, I then match for repo name.  This isn't really a security concern, as only our repo has access to dev deployment variables and only your repo has access to staging/prod vars.

File: README.md [hunk 2]
```diff
@@ -35,30 +35,60 @@ You may run into some issues running the docker commands on Windows:
...
+| | Run `yarn lint:ci` for both the frontend and backend | `yarn lint:all`| |
+
+Integration
+-----------
+
+Linting, unit tests and test coverage analysis are all run automatically on each push
+to the Ad Hoc fork of HHS/Head-Start-TTADP repo and the HHS/Head-Start-TTADP repo. In
+the Ad Hoc repository, merges to the main branch are blocked if the continuous
+integration (CI) tests do not pass. The continuous integration pipeline is configured via CircleCi.
+The bulk of CI configurations can be found in this repo's [.circleci/config.yml](.circleci/config.yml) file.
```
Comment (SarahJaine) [inline README.md:64]: I'll update this section to mention the security and accessibility reviews once those are in place on the CI pipeline.

File: README.md [hunk 3]
```diff
@@ -35,30 +35,60 @@ You may run into some issues running the docker commands on Windows:
...
+| | Run tests for the backend with coverage and output results to xml files|  `yarn test:ci`| |
+| | Run `yarn test:ci` for both the frontend and backend | `yarn test:all`| |
+| | Run the linter only for the backend | `yarn lint` | |
+| | Run the linter for the the backend with results output to xml files | `yarn lint:ci`| |
+| | Run `yarn lint:ci` for both the frontend and backend | `yarn lint:all`| |
+
+Integration
+-----------
+
+Linting, unit tests and test coverage analysis are all run automatically on each push
```
Comment (SarahJaine) [inline README.md:60]: I'll update this section to mention the security and accessibility reviews once those are in place on the CI pipeline.

File: .circleci/config.yml [hunk 4]
```diff
@@ -25,6 +25,70 @@ commands:
...
           name: Combine package-lock.json files to single file
           command: cat yarn.lock frontend/yarn.lock > << parameters.filename >>
+  cf_deploy:
+    description: "Login to cloud foundry space with service account credentials
+      and push application using deployment configuration file."
+    parameters:
+      cloudgov_username:
+        description: "Name of CircleCi project environment variable that
+          holds deployer username for cloudgov space"
+        type: env_var_name
```
Comment (SarahJaine) [inline .circleci/config.yml:35]: I tried echoing out a parameter that holds an env var, and circle ci still knows to redact it from the logs.

**Review by rahearn — CHANGES_REQUESTED** [review]
Awesome work. In addition to the change in routes, I'm wondering if the deploy step needs to set `NODE_ENV=production` for the two `yarn build` steps. Does build do things different for production vs development? And is that how it determines what to build?

File: .circleci/config.yml [hunk 5]
```diff
@@ -84,16 +148,65 @@ jobs:
...
+          name: Build backend assets
+          command: yarn build
+      - run:
+          name: Build frontend assets
+          command: cd frontend && yarn build
+      - run:
+          name: Install Cloud Foundry
+          command: |
+            curl -v -L -o cf-cli_amd64.deb 'https://cli.run.pivotal.io/stable?release=debian64&source=github'
+            sudo dpkg -i cf-cli_amd64.deb
```
[the remark this answers is not in the document]
Comment (SarahJaine) [reply .circleci/config.yml:171]: Thanks! I was interested to see your feedback on this line. I thought I might be missing something.

File: .circleci/config.yml [hunk 6]
```diff
@@ -84,16 +148,65 @@ jobs:
...
+    executor: docker-executor
+    steps:
+      - attach_workspace:
+          at: .
+      - run:
+          name: Build backend assets
+          command: yarn build
+      - run:
+          name: Build frontend assets
+          command: cd frontend && yarn build
```
[the remark this answers is not in the document]
Comment (SarahJaine) [reply .circleci/config.yml:166]: I could have _sworn_ I changed all these. I think this change got lost in a bad merge from upstream.  I'll fix.

File: README.md [hunk 7]
```diff
@@ -35,30 +35,60 @@ You may run into some issues running the docker commands on Windows:
...
+
+* **Dev** deployed by commit to [main branch][adhoc-main] of Ad Hoc fork of HHS/Head-Start-TTADP repo
+* **Staging** deployed by commit to [main branch][hhs-main] of HHS/Head-Start-TTADP repo
+* **Prod** deployed by commit to [production branch][hhs-prod] of HHS/Head-Start-TTADP repo
+
+The application is deployed to the following URLs:
+
+| Environment | URL |
+|:------------|:----|
+| dev     | https://tta-smarthub.app.cloud.gov/ |
```
[the remark this answers is not in the document]
Comment (rahearn) [reply README.md:85]: The route has to be globally unique, while app names are fine to be duplicated in different spaces.

I'm fine with your other method of setting a unique name and getting the route from that, but I'd like to get either that whole PR or just that change included in this PR or the deploy step will fail here.

**Comment by SarahJaine** [comment]
> Awesome work. In addition to the change in routes, I'm wondering if the deploy step needs to set `NODE_ENV=production` for the two `yarn build` steps. Does build do things different for production vs development? And is that how it determines what to build?

Are you asking how/if I created an application build script specific to the deployed envs? For deployment, I'm using a few new commands (added this PR):
- for BE - `yarn build` uses babel to compile the files and puts them in a `build/server` dir. `yarn start` then runs that new compiled version of index.js to start the BE server.
- for FE - `yarn build` still  uses react-scripts like it does for local development, but those files are moved to a `build/server/client` dir.  I updated index.js so it can find those static files.

That setup is what's allowing me to not push (via .cfignore) our `src` dir (our normal non-compiled BE) and `frontend` dir (our normal non-compiled FE) onto the cf instance when it's deployed.

Does that help? Did I miss your question?

File: .circleci/config.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (SarahJaine) [reply .circleci/config.yml:166]: ✅  Added in https://github.com/adhocteam/Head-Start-TTADP/pull/29

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (SarahJaine) [reply README.md:85]: Gotcha, moving this conversation to slack to figure out a merge strategy that doesn't make everyone crazy.

**Comment by rahearn** [comment]
> * for BE - `yarn build` uses babel to compile the files

My comment is related to using something like `NODE_ENV=production yarn build` because sometimes (though it doesn't look like this project is currently set up that way) different babel plugins are configured for development vs production. Things like tree-shaking the front end in production, but not in development.

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (rahearn) [reply README.md:85]: Agreed here: https://gsa-tts.slack.com/archives/C017RJ4GEET/p1600453033043500?thread_ts=1600452758.040600&cid=C017RJ4GEET to make this change as part of another PR.

**Comment by SarahJaine** [comment]
> > * for BE - `yarn build` uses babel to compile the files
> 
> My comment is related to using something like `NODE_ENV=production yarn build` because sometimes (though it doesn't look like this project is currently set up that way) different babel plugins are configured for development vs production. Things like tree-shaking the front end in production, but not in development.

We didn't add any configuration to the make the local/development and deployed/production builds of the babel plugins different.  To be honest, I haven't seen that before and didn't know it was a thing.  Do you want me to look into doing that as part of this PR?

**Comment by rahearn** [comment]
> Do you want me to look into doing that as part of this PR?

nope, it might turn into a story in the future, but if the builds are the same for both environments right now we can leave it as be. Thanks

**Comment by rahearn** [comment]
Definition of Done checklist:

- [x] The business and/or user value of a story has been met
- [x] CI pipeline green
- [x] Code is meaningfully tested
- [x] No linting errors or warnings
- [x] UI meets WCAG 2.1 - Levels A and AA
- [x] Code has deployed to a staging environment
- [x] [Architectural Decision Records](https://adr.github.io/) are written for major infrastructure decisions with [the Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md).
- [x] Boundary and Data Flow Diagrams are updated when a change invalidates the current version.
- [x] Public API methods are documented when added or changed.
- [x] OSCAL templates are completed when security controls are implemented or modified.

**Review by rahearn — APPROVED** [review]

**Pull request merged**